internal/staging/collector.go

```diff
@@ -0,0 +1,230 @@
+	"k8s.io/apimachinery/pkg/types"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("staging")
+
+var (
+	eventsCollected = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "staging_events_collected_total",
+		Help: "Total number of reconciliation events collected",
+	})
+
+	batchSizeHistogram = promauto.NewHistogram(prometheus.HistogramOpts{
+		Name:    "staging_batch_size",
+		Help:    "Size of reconciliation batches processed",
+		Buckets: prometheus.LinearBuckets(1, 10, 20), // 1-200 with step 10
+	})
+)
+
+type EventType string
+
+const (
+	EventTypeCreate EventType = "Create"
+	EventTypeUpdate EventType = "Update"
+	EventTypeDelete EventType = "Delete"
+)
+
+type EventInfo struct {
+	Resource        rules.WSAResource
+	EventType       EventType
+	Generation      int64
+	ResourceVersion string
+	Timestamp       time.Time
+}
+
+type EventCollector struct {
+	eventMap         map[types.NamespacedName]*EventInfo
+	debounceInterval time.Duration
+	maxBatchSize     int
+
+	mu           sync.RWMutex
+	batchReadyCh chan []*EventInfo
+}
+
+func NewEventCollector(debounceInterval time.Duration, maxBatchSize int) *EventCollector {
+	return &EventCollector{
+		eventMap:         make(map[types.NamespacedName]*EventInfo),
+		debounceInterval: debounceInterval,
+		maxBatchSize:     maxBatchSize,
+		batchReadyCh:     make(chan []*EventInfo, 10),
+	}
+}
+
+func (ec *EventCollector) AddEvent(event *EventInfo) {
+	ec.mu.Lock()
+
+	key := types.NamespacedName{
+		Namespace: event.Resource.GetNamespace(),
+		Name:      event.Resource.GetName(),
+	}
+
+	ec.eventMap[key] = event
+	eventsCollected.Inc()
+
+	shouldTrigger := len(ec.eventMap) >= ec.maxBatchSize
+	if shouldTrigger {
+		batch := ec.prepareBatch()
+		ec.mu.Unlock()
+		ec.sendBatch(batch)
+	} else {
+		ec.mu.Unlock()
+	}
+}
+
+func (ec *EventCollector) Start(ctx context.Context) error {
+	ticker := time.NewTicker(ec.debounceInterval)
+	defer ticker.Stop()
+
+	log.Info("EventCollector started", "debounceInterval", ec.debounceInterval, "maxBatchSize", ec.maxBatchSize)
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(ec.batchReadyCh)
+			log.Info("EventCollector stopped")
+			return nil
+		case <-ticker.C:
+			ec.mu.Lock()
+			var batch []*EventInfo
+			if len(ec.eventMap) > 0 {
+				batch = ec.prepareBatch()
+			}
+			ec.mu.Unlock()
+
+			if batch != nil {
+				ec.sendBatch(batch)
+			}
+		}
+	}
+}
+
+// prepareBatch extracts events from eventMap and clears the map.
+// Caller must hold ec.mu.Lock()
+func (ec *EventCollector) prepareBatch() []*EventInfo {
+	batch := make([]*EventInfo, 0, len(ec.eventMap))
+	for _, event := range ec.eventMap {
+		batch = append(batch, event)
+	}
+
+	ec.eventMap = make(map[types.NamespacedName]*EventInfo)
+	return batch
+}
+
+// Flush forces any pending events to be sent as a batch immediately.
+// Returns true if a batch was sent, false if there were no pending events.
+func (ec *EventCollector) Flush() bool {
```

Comment:
This logic is used in a few places (eg ln 101) - let's commonise it.
The return from `Flush` is not used